test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,325 @@
+end
+
+-- Test that box_promote will return 0
+-- if server is already raft leader and limbo owner.
+g.test_leader_promote = function(g)
+    local ok, _ = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(ok, 'error while promoting leader')
+end
+
+-- Test that box_promote will return 0
+-- if server is already raft leader but not current limbo owner.
+g.test_raft_leader_promote = function(g)
+    g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_BOX_RAFT_SYNCHRO_QUEUE_DELAY', true)
+        box.ctl.promote()
+    end)
+    g.server_1:wait_election_state('leader')
+
+    local ok = g.server_1:exec(function()
+        local ok, _ = pcall(box.ctl.promote)
+        box.error.injection.set('ERRINJ_BOX_RAFT_SYNCHRO_QUEUE_DELAY', false)
+        return ok
+    end)
+    luatest.assert(ok, 'error while promoting raft leader')
+end
+
+-- Test that box_demote will return 0
+-- if server is already follower.
+g.test_follower_demote = function(g)
+    local ok, _ = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(ok, 'error while demoting follower')
+end
+
+-- Test that box_promote will return box.error.UNSUPPORTED
+-- when called while in promote already
+g.test_simultaneous_promote = function(g)
+    local ok, err = g.server_1:exec(function()
+        require('fiber').create(box.ctl.promote)
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.UNSUPPORTED,
+        'error while promoting while in promote')
+end
+
+-- Test that box_demote will return box.error.UNSUPPORTED
+-- when called while in promote already
+g.test_simultaneous_demote = function(g)
+    local ok, err = g.server_1:exec(function()
+        require('fiber').create(box.ctl.promote)
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.UNSUPPORTED,
+        'error while demoting while in promote')
+end
+
+-- Test that box_promote will return box.error.UNSUPPORTED
+-- when trying to promote voter
+g.test_voter_promote = function(g)
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(
+        not ok and err.code == box.error.UNSUPPORTED,
+        'error while promoting voter')
+end
+
+-- Test interfering promotion for election_mode = 'off'
+-- while in WAL delay
+g.test_wal_interfering_promote = function(g)
+    local wal_write_count = g.server_1:exec(function()
+        box.error.injection.set('ERRINJ_WAL_DELAY', true)
+        return box.error.injection.get('ERRINJ_WAL_WRITE_COUNT')
+    end)
+
+    g.server_2:exec(function()
+        box.ctl.promote()
+    end)
+    g.server_2:wait_synchro_queue_owner()
+    g.server_1:wait_wal_write_count(wal_write_count + 1)
+
+    local ok, err = g.server_1:exec(function()
+        local f = require('fiber').new(box.ctl.promote); f:set_joinable(true)
```

Comment:
`fiber.new()` does not yield. It means that you will set the delay to `false` below before `box.ctl.promote()` is called in your new fiber. That in turn raises 2 questions - why do you need to start the promote before turning off the injection and are you sure the test works as intended then?